### ISSUE TITLE
Files removed from project stay in index

### ISSUE THREAD

**Issue by ckandeler** [clangd/clangd#935]
Consider this sequence of events:

- Start clangd with a compilation db containing a source file source.cpp.
- Wait for project to get indexed.
- Do textDocument/references on some given symbol. 
- Stop clangd.
- Rename source.cpp to source2.cpp and update compilation db file accordingly.
- Start clangd with the new compilation db and wait for indexing to finish.
- Do textDocument/references on the same symbol. Now clangd will find it in two files: source.cpp and source2.cpp. The former does not even exist anymore (but if it did, it should still not be in the index).

Output of `clangd --version`:
clangd version 14.0.0 (https://github.com/llvm/llvm-project.git 9f0b5f9a39ea6e70c98c69a720d7e4f5d3800bf6)
Features: linux
Platform: x86_64-unknown-linux-gnu

Editor/LSP plugin:
Qt Creator

**Comment by sam-mccall** [comment]
I can't reproduce using the steps you gave, and I can't see how this would happen, but it might in more complex setups.

Clangd doesn't scan/load all shards in .cache/clangd/index; they're loaded only when referenced.
 - when compile_commands.json is found, all the filenames are put on a queue
 - for each filename in the queue, we try to load its index shard
 - if the shard is loaded, we add all the filenames of its included headers to the queue (these are stored in the shard)

In the scenario you describe, clangd should never touch `source.cpp.*.idx` when started for the second time, and that's what I see.

However if it was a header file being renamed, things would be different: `source.h` could be mentioned in `main.cpp.*.idx`, so `source.h.*.idx` would be loaded. This is true even if `main.cpp` only mentions `source2.h`: we load even stale shards and their dependencies (and would then enqueue `main.cpp` for reindexing).

Could this be what you're seeing?

**Comment by ckandeler** [comment]
> However if it was a header file being renamed, things would be different:

Yes, it is actually a header file. Apologies for over-simplifying the case. The originating bug report (with example project) is here: https://bugreports.qt.io/browse/QTCREATORBUG-26574.


> This is true even if main.cpp only mentions source2.h: we load even stale shards and their dependencies (and would then enqueue main.cpp for reindexing

This sounds like the stale entry should be gone after this re-indexing?

**Comment by sam-mccall** [comment]
No worries, thanks for the link.

> This sounds like the stale entry should be gone after this re-indexing?

The references that would cause the zombie shard to be loaded are gone, but nothing actually unloads it from memory.
So it'll stick around until you restart clangd *again* :-(

Think of the background index as `{ queue<string> FilesToProcess; map<string, Symbols> FileSymbols; }`. As files are processed (loaded/indexed), entries in FileSymbols get added/updated, but there's no refcounting that would remove an entry.

All this is pretty sad. I can see a few options:
 - don't load the header shard at all if the file doesn't exist. (We already check for staleness). In the rename case we now lose the symbols entirely until reindexing instead of thinking they're in the wrong header.
 - refcount the shards and remove them if a header is no longer reachable from any file listed in the CDB. This seems maybe complicated.
 - periodically scan files on disk to check for staleness, and evict missing ones. This has a delay and also stops us going truly idle, but having the index update when disk changes would be nice to have too.

The first option seems like the winner to me at least for now.

**Comment by sam-mccall** [comment]
(This is all separate from the fact that the file stays around on disk forever. Maybe we should remove it if the file doesn't exist, but I'm a bit worried that's a lot of churn if the user is switching branches or so. So I'm inclined not to try cleaning up the files)
